[tool] cleanup bundle builder a bit

ardera opened the pull request:
- remove buildWithAssemble shim since only BundleBuilder was using it
- move that functionality into the BundleBuilder itself
- source all the build parameters and build system defines from buildInfo instead of extra parameters
- make the bundle_shim_test.dart test the BundleBuilder instead (and rename it)
- migrate all the tests and usages (that's only the flutter build bundle command and CustomDevices)
- added one new test that makes sure all the right environment defines are passed to the build system

fixes #80732 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

ardera committed:
remove buildWithAssemble shim, since only BundleBuilder was using it
- move that functionality into the BundleBuilder itself
- source all the build parameters and build system defines from buildInfo instead of extra parameters
- make the bundle_shim_test.dart test the BundleBuilder instead (and rename it)
- migrate all the tests and usages (that's only the flutter build bundle command and CustomDevices)

ardera committed:
add bundle builder test that makes sure all the right Environment defines are passed to the build system

jonahwilliams commented on packages/flutter_tools/lib/src/bundle.dart:
```diff
@@ -84,34 +84,83 @@ class BundleBuilder {
   /// The default  `manifestPath` is `pubspec.yaml`
   Future<void> build({
```
You could arrange this a little bit differently to make the unit tests a bit easier to write for now (avoiding the top level `env` declaration.

Add a new optional parameter: `@visibleForTesting BuildSystem buildSystemOverride`

Then use it with something like `buildSystemOverride ?? globals.buildSystem`. You can then pass the build system directly to the bundle builder

ardera replied:
done

ardera committed:
allow overriding BuildSystem in BundleBuilder.build for testing
- add new tests for flutter build bundle command (since things like filesystem scheme, roots, frontend & gen_snapshot options weren't passed correctly before and no test was failing)
- fix the old flutter build bundle tests
- only add extraFrontEndOptions, extraGenSnapshotOptions and fileSystemRoots to the build env when they're not empty lists

ardera committed:
Update bundle_builder_test.dart

ardera commented:
I think the one failing test is a fluke, only moved the `env` variable into the `testUsingContext` closure and it worked fine before that

jonahwilliams commented:
looks like a timeout rather than a flake

jonahwilliams commented:
hmm but its an integration test, I'll give it a kick

ardera committed:
make the other tests use the BundleBuilder().build buildsystem argument as well

ardera committed:
Merge remote-tracking branch 'origin/feature/bundle-building-refactor' into feature/bundle-building-refactor

ardera commented on packages/flutter_tools/test/general.shard/bundle_builder_test.dart:
```diff
@@ -0,0 +1,123 @@
+// Tests for BundleBuilder.
+void main() {
+  testUsingContext('Copies assets to expected directory after building', () async {
+    await BundleBuilder().build(
```
yep forgot about that for this test, however I'm not sure I can save any overrides, `BundleBuilder.build` accesses some globals directly

Generally, what stuff should be passed via the app context and what via arguments? Are you trying to use arguments as much as possible, or environment or does it not matter? I mean theoretically I could also make `BundleBuilder` accept the `fileSystem`, `cache`, `processManager`, `buildContext` and so on via the constructor similiar to how all the Device-related (`CustomDevice` etc) classes do it

jonahwilliams replied:
global context injection is the legacy method of providing dependencies. Generally we've been slowly, slowly migrating code to just use regular old constructor injection of deps, or to provide visibleForTesting arguments where globals are still used heavily. If you're refactoring or rewriting code, prefer restructuring it to not depend on context injection

ardera replied:
Ah okay. Yeah I might as well use constructor arguments then

jonahwilliams reviewed (approved):
LGTM

ardera commented:
nice!

> looks like a timeout rather than a flake

oh, thought thats the same thing
